File: utils/MeshReader/Colormap.cpp
#pragma once
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>
#include <map>

/*
Classe de leitura de arquivos .mtl, que guarda cores e propriedades de materiais.

A saber que:
    - kd = Difuso (Cor do objeto)
    - ks = Specular (Reflexivo)
    - ke = Emissivo 
    - ka = Ambiente
    - ns = Brilho
    - ni = Índice de refração
    - d = Opacidade

A classe precisa ser instânciada passando o caminho do arquivo .mtl correspondente
*/


using namespace std;

struct MaterialProperties {
    Vetor  kd;     // Difuso      (kd)
    Vetor  ks;     // Especular   (ks)
    Vetor  ka;     // Ambiente    (ka)
    Vetor  kr;     // Reflexivo   (kr)  [Emissivo  (ke)] considere ke = kr
    Vetor  ke;     // Emissivo    (ke)  [Reflexivo (kr)] considere kr = ke  //normalmente o ke é o mais usado
    Vetor  kt;     // Transmicivo (kt)
    double ns;     // Rugosidade  (eta) [Brilho]
    double ni;     // Refração 
    double d;      // Opacidade

    MaterialProperties() : kd(0, 0, 0), ks(0, 0, 0), ke(0, 0, 0), ka(0, 0, 0), ns(0), ni(0), d(0) {}
};

class colormap {

public:
    map<string, MaterialProperties> mp;

    //Construtor    
    colormap(){};
    colormap(string input){

        // construtor: lê arquivo cores.mtl e guarda valores RGB associados a cada nome

        std::ifstream mtlFile(input);

        if (!mtlFile.is_open()) {
            std::cerr << "erro abrindo arquivo cores.mtl\n";
        }

        string line, currentMaterial;

        while (std::getline(mtlFile, line)) {
            std::istringstream iss(line);
            std::string keyword;
            iss >> keyword;

            if (keyword == "newmtl") {
                iss >> currentMaterial;
                if (!currentMaterial.empty()) {
                    mp[currentMaterial] = MaterialProperties();
                }
            } else if (keyword == "Kd") {
                double kdR, kdG, kdB;
                iss >> kdR >> kdG >> kdB;
                if (!currentMaterial.empty()) {
                    mp[currentMaterial].kd = Vetor(kdR, kdG, kdB);
                }
            } else if (keyword == "Ks") {
                double ksR, ksG, ksB;
                iss >> ksR >> ksG >> ksB;
                if (!currentMaterial.empty()) {
                    mp[currentMaterial].ks = Vetor(ksR, ksG, ksB);
                }
            } else if (keyword == "Ke") {
                double keR, keG, keB;
                iss >> keR >> keG >> keB;
                if (!currentMaterial.empty()) {
                    mp[currentMaterial].ke = Vetor(keR, keG, keB);
                }
            } else if (keyword == "Kr") {
                double krR, krG, krB;
                iss >> krR >> krG >> krB;
                if (!currentMaterial.empty()) {
                    mp[currentMaterial].kr = Vetor(krR, krG, krB);
                }
            }
            else if (keyword == "Ka") {
                double kaR, kaG, kaB;
                iss >> kaR >> kaG >> kaB;
                if (!currentMaterial.empty()) {
                    mp[currentMaterial].ka = Vetor(kaR, kaG, kaB);
                }
            } else if (keyword == "Ns") {
                iss >> mp[currentMaterial].ns;
            } else if (keyword == "Ni") {
                iss >> mp[currentMaterial].ni;
            } else if (keyword == "d") {
                iss >> mp[currentMaterial].d;
            }
        }

        mtlFile.close();
    }

    Vetor getColor(string& s){
        if (mp.find(s) != mp.end()) {
            return mp[s].kd;
        } else {
            cerr << "Error: cor " << s << " indefinida no arquivo .mtl\n";
            return Vetor(0,0,0);
        }
    }

    MaterialProperties getMaterialProperties(string& s){
        if (mp.find(s) != mp.end()) {
            return mp[s];
        } else {
            cerr << "Error: Cor " << s << " indefinida no arquivo .mtl\n";
            return MaterialProperties();
        }
    }

};

File: utils/MeshReader/Colormap.py
"""
Leitor de arquivos .mtl (Material Template Library).

Propriedades de material:
    kd = Difuso (cor do objeto)
    ks = Especular (reflexivo)
    ke = Emissivo
    ka = Ambiente
    kr = Reflexivo  (considere kr = ke)
    kt = Transmissivo
    ns = Brilho / rugosidade
    ni = Índice de refração
    d  = Opacidade
"""
from __future__ import annotations
import sys
from dataclasses import dataclass, field
from pathlib import Path

from src.Vetor import Vetor


@dataclass
class MaterialProperties:
    kd: Vetor = field(default_factory=lambda: Vetor(0, 0, 0))  # Difuso
    ks: Vetor = field(default_factory=lambda: Vetor(0, 0, 0))  # Especular
    ka: Vetor = field(default_factory=lambda: Vetor(0, 0, 0))  # Ambiente
    kr: Vetor = field(default_factory=lambda: Vetor(0, 0, 0))  # Reflexivo  (alias ke)
    ke: Vetor = field(default_factory=lambda: Vetor(0, 0, 0))  # Emissivo   (alias kr)
    kt: Vetor = field(default_factory=lambda: Vetor(0, 0, 0))  # Transmissivo
    ns: float = 0.0   # Brilho / rugosidade
    ni: float = 0.0   # Índice de refração
    d:  float = 0.0   # Opacidade


class Colormap:
    """Lê um arquivo .mtl e expõe as propriedades de cada material por nome."""

    def __init__(self, filepath: str = ""):
        self.mp: dict[str, MaterialProperties] = {}
        if not filepath:
            return

        path = Path(filepath)
        if not path.exists():
            print(f"erro abrindo arquivo {filepath}", file=sys.stderr)
            return

        current_material = ""
        with open(path, "r", encoding="utf-8") as f:
            for line in f:
                parts = line.strip().split()
                if not parts:
                    continue
                keyword, *rest = parts

                if keyword == "newmtl":
                    current_material = rest[0] if rest else ""
                    if current_material:
                        self.mp[current_material] = MaterialProperties()

                elif current_material:
                    mat = self.mp[current_material]
                    if keyword == "Kd" and len(rest) >= 3:
                        mat.kd = Vetor(float(rest[0]), float(rest[1]), float(rest[2]))
                    elif keyword == "Ks" and len(rest) >= 3:
                        mat.ks = Vetor(float(rest[0]), float(rest[1]), float(rest[2]))
                    elif keyword == "Ke" and len(rest) >= 3:
                        mat.ke = Vetor(float(rest[0]), float(rest[1]), float(rest[2]))
                    elif keyword == "Kr" and len(rest) >= 3:
                        mat.kr = Vetor(float(rest[0]), float(rest[1]), float(rest[2]))
                    elif keyword == "Ka" and len(rest) >= 3:
                        mat.ka = Vetor(float(rest[0]), float(rest[1]), float(rest[2]))
                    elif keyword == "Ns" and rest:
                        mat.ns = float(rest[0])
                    elif keyword == "Ni" and rest:
                        mat.ni = float(rest[0])
                    elif keyword == "d" and rest:
                        mat.d = float(rest[0])

    def get_color(self, name: str) -> Vetor:
        if name in self.mp:
            return self.mp[name].kd
        print(f"Error: cor {name!r} indefinida no arquivo .mtl", file=sys.stderr)
        return Vetor(0, 0, 0)

    def get_material_properties(self, name: str) -> MaterialProperties:
        if name in self.mp:
            return self.mp[name]
        print(f"Error: material {name!r} indefinido no arquivo .mtl", file=sys.stderr)
        return MaterialProperties()
